pkg/kubelet/network/hostport/hostport.go

```diff
@@ -140,32 +145,127 @@ func portMappingToHostport(portMapping *PortMapping) hostport {
-	args := []string{"-m", "comment", "--comment", "kube hostport portals",
+	args := []string{
+		"-m", "comment", "--comment", "maybe kube hostport",
 		"-m", "addrtype", "--dst-type", "LOCAL",
-		"-j", string(kubeHostportsChain)}
+		"-j", string(kubeLocalDestChain),
+	}
 	for _, tc := range tableChainsNeedJumpServices {
 		if _, err := iptables.EnsureRule(utiliptables.Prepend, tc.table, tc.chain, args...); err != nil {
 			return fmt.Errorf("Failed to ensure that %s chain %s jumps to %s: %v", tc.table, tc.chain, kubeHostportsChain, err)
 		}
+		removeOldLink(iptables, tc.table, tc.chain) // ignore errors
+	}
+	removeOldSNAT(iptables) // ignore errors
+
+	return nil
+}
+
+func syncLocaldest(iptables utiliptables.Interface) error {
+	start := time.Now()
+	defer func() {
+		glog.V(4).Infof("syncLocaldest took %v", time.Since(start))
+	}()
+
+	// Get iptables-save output so we can check for existing chains and rules.
+	// This will be a map of chain name to chain with rules as stored in iptables-save/iptables-restore
+	existingNATChains := make(map[utiliptables.Chain]string)
+	iptablesSaveRaw, err := iptables.Save(utiliptables.TableNAT)
+	if err != nil { // if we failed to get any rules
+		glog.Errorf("Failed to execute iptables-save, syncing all rules: %v", err)
+	} else { // otherwise parse the output
+		existingNATChains = utiliptables.GetChainLines(utiliptables.TableNAT, iptablesSaveRaw)
+	}
+
+	natChains := bytes.NewBuffer(nil)
+	natRules := bytes.NewBuffer(nil)
+	writeLine(natChains, "*nat")
+	// Make sure we keep stats for the top-level chains, if they existed
+	// (which most should have because we created them above).
+	if chain, ok := existingNATChains[kubeLocalDestChain]; ok {
+		writeLine(natChains, chain)
+	} else {
+		writeLine(natChains, utiliptables.MakeChainLine(kubeLocalDestChain))
+	}
+
+	// Populate the LOCALDEST chain.
+	addrs, err := net.InterfaceAddrs()
+	if err != nil {
+		return err
 	}
-	// Need to SNAT traffic from localhost
-	args = []string{"-m", "comment", "--comment", "SNAT for localhost access to hostports", "-o", natInterfaceName, "-s", "127.0.0.0/8", "-j", "MASQUERADE"}
-	if _, err := iptables.EnsureRule(utiliptables.Append, utiliptables.TableNAT, utiliptables.ChainPostrouting, args...); err != nil {
-		return fmt.Errorf("Failed to ensure that %s chain %s jumps to MASQUERADE: %v", utiliptables.TableNAT, utiliptables.ChainPostrouting, err)
+	// Only consider packets with dest of a real interface addr for hostports.
+	for _, a := range addrs {
+		astr := a.String() // CIDR notation
+		ip, ipnet, err := net.ParseCIDR(astr)
+		if err != nil {
+			return err
+		}
+		if ip.To4() != nil {
+			ipstr := ip.String()
+			if ip.IsLoopback() {
+				// Special case loopback devices.
+				ipstr = ipnet.String()
+			}
+			args := []string{
+				"-A", string(kubeLocalDestChain),
+				"-m", "comment", "--comment", `"maybe kube hostport"`,
+				"-d", ipstr,
+				"-j", string(kubeHostportsChain),
+			}
+			writeLine(natRules, args...)
```

Comment:
So this code gets run both at HostportManager creation, but also whenever a new hostport is added.  That means if interfaces come and go (which is not uncommon), especially if they are something like libvirt interfaces from VMs, you'll build up a list of rules that you'll never clean up.

I think you need to save the existing output, and then -X any rules from that output that aren't found in the this loop here.